File: kotlin-app/composeApp/src/commonMain/kotlin/app/sufrix/ui/AmountField.kt
package app.sufrix.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.wrapContentHeight
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.focus.onFocusChanged
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlin.math.roundToLong

// A big, tabular money input — the opening-cash / tender hero field. Edits a
// major-unit decimal string and reports `amountMinor` (minor units). Mirror of
// the SwiftUI AmountField.
@Composable
fun AmountField(
    amountMinor: Long,
    onAmountMinor: (Long) -> Unit,
    currencyCode: String,
    modifier: Modifier = Modifier,
    autofocus: Boolean = false,
) {
    val c = sufrixColors()
    var text by remember { mutableStateOf(if (amountMinor == 0L) "" else minorToText(amountMinor)) }
    // The last value WE emitted — lets us tell an external change (a carried-over
    // prefill arriving async) from the teller's own typing, with no focus races.
    var lastEmitted by remember { mutableStateOf(amountMinor) }
    var focused by remember { mutableStateOf(false) }
    val focusRequester = remember { FocusRequester() }
    LaunchedEffect(autofocus) {
        if (autofocus) runCatching { focusRequester.requestFocus() }
    }
    LaunchedEffect(amountMinor) {
        if (amountMinor != lastEmitted) {
            text = if (amountMinor == 0L) "" else minorToText(amountMinor)
            lastEmitted = amountMinor
        }
    }

    // One contained row: a muted currency prefix on the left, the big tabular
    // amount filling the rest (was a tiny label stacked over a 34sp number).
    Row(
        modifier
            .fillMaxWidth()
            .height(64.dp)
            .clip(RoundedCornerShape(Radii.md))
            .background(c.surface)
            .border(
                if (focused) 2.dp else 1.dp,
                if (focused) c.accent else c.border,
                RoundedCornerShape(Radii.md),
            )
            .padding(horizontal = Space.lg),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(Space.sm),
    ) {
        Text(
            currencyCode.uppercase(),
            color = c.textMuted, fontFamily = SufrixFont,
            fontWeight = FontWeight.Bold, fontSize = 15.sp,
        )
        Box(Modifier.weight(1f).wrapContentHeight()) {
            if (text.isEmpty()) {
                Text(
                    "0.00", color = c.textMuted, fontFamily = SufrixFont,
                    fontWeight = FontWeight.Black, fontSize = 28.sp,
                )
            }
            BasicTextField(
                value = text,
                onValueChange = { newValue ->
                    text = newValue
                    val m = toMinor(newValue)
                    lastEmitted = m
                    onAmountMinor(m)
                },
                singleLine = true,
                modifier = Modifier.fillMaxWidth().focusRequester(focusRequester).onFocusChanged { focused = it.isFocused },
                textStyle = TextStyle(
                    color = c.textPrimary, fontFamily = SufrixFont,
                    fontWeight = FontWeight.Black, fontSize = 28.sp,
                ),
                keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Decimal),
                cursorBrush = SolidColor(c.accent),
            )
        }
    }
}

/** Parse a major-unit decimal string ("500", "499.50") → minor units. */
private fun toMinor(s: String): Long {
    val cleaned = s.filter { it.isDigit() || it == '.' }
    val major = cleaned.toDoubleOrNull() ?: 0.0
    return (major * 100).roundToLong()
}

private fun minorToText(minor: Long): String {
    val major = minor / 100.0
    return if (major == major.toLong().toDouble()) major.toLong().toString()
    else ((major * 100).roundToLong() / 100.0).toString()
}

File: swift-app/Sources/MadarUI/Components/AmountField.swift
// A big, tabular money input — the opening-cash / tender hero field. Edits a
// major-unit decimal string and reports `amountMinor` (minor units).
import SwiftUI

struct AmountField: View {
    @Environment(\.theme) private var theme
    @Environment(\.localize) private var t
    @FocusState private var focused: Bool

    @Binding var amountMinor: Int64
    var currencyCode: String
    /// Raise the keyboard on appear (the hero count field; off by default so
    /// existing call sites are unchanged).
    var autofocus: Bool = false

    @State private var text = ""
    /// The last value WE emitted — distinguishes an external prefill from the
    /// teller's own typing, so prefills reflect without a focus race.
    @State private var lastEmitted: Int64 = 0

    var body: some View {
        // One contained row: a muted currency prefix on the left, the big tabular
        // amount filling the rest. (Was a tiny label stacked over a 34pt number,
        // which sprawled and read as unbalanced.)
        HStack(spacing: Space.sm) {
            Text(currencyCode.uppercased())
                .font(.ui(15, .bold))
                .foregroundStyle(theme.colors.textMuted)
            TextField("0.00", text: Binding(
                get: { text },
                set: { newValue in
                    text = newValue
                    let minor = Self.toMinor(newValue)
                    lastEmitted = minor
                    amountMinor = minor
                }
            ))
            .font(.money(28, .heavy))
            .textFieldStyle(.plain) // suppress macOS's default rounded-border bezel
            .foregroundStyle(theme.colors.textPrimary)
            .focused($focused)
            #if os(iOS)
            .keyboardType(.decimalPad)
            // The decimalPad has no return key — give it a Done dismiss.
            .toolbar {
                ToolbarItemGroup(placement: .keyboard) {
                    Spacer()
                    Button(t("common.done")) { focused = false }
                }
            }
            #endif
        }
        .padding(.horizontal, Space.lg)
        .frame(height: 64)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(theme.colors.surface)
        .overlay(
            RoundedRectangle(cornerRadius: Radii.md, style: .continuous)
                .strokeBorder(focused ? theme.colors.accent : theme.colors.border, lineWidth: focused ? 2 : 1)
        )
        .clipShape(RoundedRectangle(cornerRadius: Radii.md, style: .continuous))
        .animation(Motion.standard, value: focused)
        .onAppear {
            // Seed the display from any pre-set amount (e.g. carried-over opening).
            lastEmitted = amountMinor
            if text.isEmpty && amountMinor != 0 { text = Self.toText(amountMinor) }
            // Let the navigation transition settle before raising the pad.
            if autofocus {
                DispatchQueue.main.asyncAfter(deadline: .now() + 0.35) { focused = true }
            }
        }
        .onChange(of: amountMinor) { newValue in
            // Reflect an EXTERNAL amount change (a carried-over prefill arriving
            // async) without clobbering the teller's own typing.
            if newValue != lastEmitted {
                text = newValue == 0 ? "" : Self.toText(newValue)
                lastEmitted = newValue
            }
        }
    }

    /// Parse a major-unit decimal string ("500", "499.50") → minor units.
    static func toMinor(_ s: String) -> Int64 {
        let cleaned = s.filter { $0.isNumber || $0 == "." }
        let major = Double(cleaned) ?? 0
        return Int64((major * 100).rounded())
    }

    /// Render minor units as a major-unit string ("48000" → "480", "49950" → "499.50").
    static func toText(_ minor: Int64) -> String {
        let major = Double(minor) / 100
        return major == major.rounded() ? String(Int64(major)) : String(format: "%.2f", major)
    }
}
